torch/csrc/dynamo/compiled_autograd_py.cpp

```diff
@@ -0,0 +1,526 @@
+  result.reserve(result_len);
+  for (const auto i : c10::irange(result_len)) {
+    result.emplace_back(THPVariable_Unpack(PyList_GET_ITEM(pyresult, i)));
+  }
+  return result;
+}
+static PyObject* check(PyObject* pyresult) {
+  if (C10_UNLIKELY(pyresult == nullptr)) {
+    // see https://github.com/pytorch/pytorch/pull/34845
+    python_error err;
+    err.persist();
+    throw err;
+  }
+  return pyresult;
+}
+
+static void check(bool result) {
+  if (C10_UNLIKELY(!result))
+    check(nullptr);
+}
+
+struct CacheNode {
+  static CacheNode* root() {
+    static CacheNode _root;
+    return &_root;
+  }
+
+  CacheNode* lookup(const CacheKey& key) {
+    auto it = next.find(key);
+    if (it == next.end()) {
+      // caller's key is in temporary memory, must copy it
+      CacheKeyBuffer buffer(key.key, key.key_size);
+      CacheKey key_with_storage(key.node_type, buffer.data, key.key_size);
+      it = next.emplace(key_with_storage, std::make_unique<CacheNode>()).first;
+      key_storage.emplace_back(std::move(buffer));
+    }
+    return it->second.get();
+  }
+
+  void clear() {
+    next.clear();
+    key_storage.clear();
+    expected_sizes.clear();
+    compiled_fn = nullptr;
+  }
+
+  bool is_empty() const {
+    return next.size() == 0 && !compiled_fn;
+  }
+
+  CacheNode() : compiled_fn(nullptr) {}
+  ~CacheNode() {
+    if (!Py_IsInitialized()) {
+      compiled_fn.release(); // leak on shutdown
+    }
+  }
+  CacheNode(CacheNode&&) = delete;
+  CacheNode(const CacheNode&) = delete;
+  CacheNode& operator=(const CacheNode&) = delete;
+  CacheNode& operator=(CacheNode&&) = delete;
+
+  bool check_dynamic_sizes(AutogradCompilerCall& call) {
+    /*
+    We start off by assuming everything is static, then we mark things
+    as dynamic when we see them change.  This function:
+      1) Checks for a cache hit
+      2) Updates expected_sizes to track what is dynamic
+      3) Populates call.dyn_size_inputs by filtering call.all_size_inputs
+    */
+    bool cache_hit = compiled_fn.get() != nullptr;
+    auto len = call.all_size_inputs.size();
+    const SizeInput* data = call.all_size_inputs.data();
+    if (expected_sizes.empty()) {
+      expected_sizes.reserve(len);
+      for (const auto i : c10::irange(len)) {
+        expected_sizes.emplace_back(data[i]);
+      }
+    }
+
+    TORCH_CHECK(expected_sizes.size() == call.all_size_inputs.size());
+    for (const auto i : c10::irange(len)) {
+      auto& expected = expected_sizes[i];
+      if (expected.dyn_type == SizeInput::DYNAMIC ||
+          expected.value != data[i].value) {
+        cache_hit = cache_hit && expected.dyn_type == SizeInput::DYNAMIC;
+        if (expected.value != data[i].value) {
+          expected = SizeInput(SizeInput::DYNAMIC, data[i].value);
+        }
+        if (call.dyn_size_inputs.empty()) {
+          call.dyn_size_inputs.reserve(len);
+        }
+        call.dyn_size_inputs.emplace_back(data[i].value);
+      }
+    }
+
+    if (!cache_hit) {
```

Comment:
// we missed cache because static size inputs didn't match; force recompilation with the varying size input as dynamic